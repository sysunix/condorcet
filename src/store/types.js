// Misc
export const RESET = "RESET";

// App
export const INITIALIZE = "INITIALIZE";
export const ADD_NOTIFCATION = "ADD_NOTIFCATION";
export const REMOVE_NOTIFICATION = "REMOVE_NOTIFICATION";
export const TOGGLE_MENU = "TOGGLE_MENU";
export const SET_REDIRECTION = "SET_REDIRECTION";
export const CLEAR_REDIRECTION = "CLEAR_REDIRECTION";

// User
export const SET_USER = "SET_USER";
export const CLEAR_USER = "CLEAR_USER";

// Poll
export const SET_POLLS = "SET_POLLS";
export const SET_POLL = "SET_POLL";
export const PUSH_POLL = "PUSH_POLL";
export const REMOVE_POLL = "REMOVE_POLL";
