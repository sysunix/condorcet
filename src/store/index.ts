import Vue from "vue";
import Vuex from "vuex";

Vue.use(Vuex);

import app from "./modules/app";
import user from "./modules/user";
import poll from "./modules/poll";

export default new Vuex.Store({
  modules: {
    app,
    user,
    poll
  }
});
