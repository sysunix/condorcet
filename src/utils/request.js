import firebase, { db } from "../firebase";

export const createPoll = async poll => {
  const pollWithTimestap = {
    ...poll,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db.collection("polls").add(pollWithTimestap);
};

export const createVote = async (vote, pollId, userId) => {
  const voteWithTimestap = {
    ...vote,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db
    .collection("polls")
    .doc(pollId)
    .collection("votes")
    .doc(userId)
    .set(voteWithTimestap);
};
