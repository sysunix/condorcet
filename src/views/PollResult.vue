<template>
  <div>
    <h3 class="text-4xl">Condorcet</h3>

    <Graph v-if="featureFlipping.graphOfTheDuels"></Graph>

    <div
      v-if="condorcet && featureFlipping.matrixOfTheDuels && answers"
      class="overflow-scroll"
    >
      <table class="text-left w-full border-collapse">
        <thead>
          <tr>
            <th
              class="py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-grey-light"
            >
              RÃ©ponse
            </th>
            <th
              v-for="answer in answers"
              :key="answer.slug"
              class="py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-grey-light"
            >
              {{ answer.value }}
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            v-for="(item, alternative) in condorcet.matrix"
            :key="alternative"
          >
            <td class="py-4 px-6 border-b border-grey-light">
              {{ item.value }}
            </td>
            <td
              v-for="opponent in item.duels"
              :key="opponent.slug"
              class="py-4 px-6 border-b border-grey-light"
            >
              {{ opponent }}
            </td>
          </tr>
        </tbody>
      </table>
    </div>

    <div v-if="condorcet && featureFlipping.condorcetRanking" class="mt-10">
      <div
        v-for="item in condorcet.ranking"
        :key="item.slug"
        class="font-sans flex items-center justify-center bg-blue-darker w-full py-2"
      >
        <div class="w-full px-6 py-4 rounded overflow-hidden shadow">
          {{ item.rank }} - <span class="font-bold">{{ item.value }}</span> /
          {{ $tc("poll.wins", item.wins, { count: item.wins }) }} &
          {{
            $tc("poll.equalities", item.equalities, { count: item.equalities })
          }}
        </div>
      </div>
    </div>

    <div v-if="uninominal && featureFlipping.uninominalRanking" class="mt-10">
      <h3 class="text-4xl">Uninominal</h3>
      <div
        v-for="item in uninominal"
        :key="item.slug"
        class="font-sans flex items-center justify-center bg-blue-darker w-full py-2"
      >
        <div class="w-full px-6 py-4 rounded overflow-hidden shadow">
          {{ item.rank }} - <span class="font-bold">{{ item.value }}</span> /
          {{
            $tc("poll.numberOfVotes", item.numberOfVotes, {
              count: item.numberOfVotes
            })
          }}
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState } from "vuex";
import sortKeys from "sort-keys";
import { db } from "../firebase";
import Graph from "../components/Graph.vue";
import { validate } from "indicative";
import { CondorcetModel, UninominalModel } from "../models/Result";

export default {
  name: "PollResult",
  components: {
    Graph
  },
  data() {
    return {
      uninominal: null,
      condorcet: null,
      answers: null
    };
  },
  computed: {
    ...mapState("app", ["featureFlipping"])
  },
  async created() {
    const resultsRef = db
      .collection("polls")
      .doc(this.$route.params.id)
      .collection("results");
    try {
      const uninominal = await validate(
        (await resultsRef.doc("uninominal").get()).data(),
        UninominalModel
      );

      this.uninominal = uninominal.ranking;

      const condorcet = await validate(
        (await resultsRef.doc("condorcet").get()).data(),
        CondorcetModel
      );

      const condorcetMatrix = Object.keys(condorcet.matrix).reduce(
        (acc, key) => ({
          ...acc,
          [key]: {
            duels: {
              ...condorcet.matrix[key].duels,
              [key]: "/"
            },
            value: condorcet.matrix[key].value
          }
        }),
        {}
      );

      this.condorcet = {
        ranking: condorcet.ranking,
        matrix: sortKeys(condorcetMatrix, { deep: true })
      };

      this.answers = condorcet.ranking
        .map(c => ({ value: c.value, slug: c.slug }))
        .sort((a, b) => {
          return a.slug.localeCompare(b.slug);
        });
    } catch (error) {
      console.log(error);
    }
  }
};
</script>

<style></style>
