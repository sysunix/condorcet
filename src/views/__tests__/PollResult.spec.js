import { shallowMount } from "@vue/test-utils";
import PollResult from "../PollResult.vue";
import store from "../../store/index";

import { MOCK_POLL_ID } from "../../utils/test";

jest.mock("../../firebase");

const $route = {
  params: { id: MOCK_POLL_ID }
};

describe("PollResult.vue", () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = shallowMount(PollResult, {
      store,
      mocks: {
        $tc: () => {},
        $route
      }
    });
  });

  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  xit("should set data", () => {
    expect(wrapper.vm.uninominal).toEqual([
      {
        rank: 1,
        value: "luffy"
      }
    ]);
  });
});
