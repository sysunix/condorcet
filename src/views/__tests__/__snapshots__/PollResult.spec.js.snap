// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollResult.vue should render 1`] = `
<div>
  <!---->
  <div>
    Merci de recharger la page
  </div>
</div>
`;
